File: datetimepickeredittext/src/main/java/io/blackbox_vision/datetimepickeredittext/view/DatePickerInputEditText.java
package io.blackbox_vision.datetimepickeredittext.view;

import android.content.Context;
import android.content.res.TypedArray;
import android.support.annotation.NonNull;
import android.support.design.widget.TextInputEditText;
import android.support.v4.app.FragmentManager;
import android.util.AttributeSet;
import android.view.View;
import android.view.inputmethod.InputMethodManager;
import android.widget.DatePicker;

import java.util.Calendar;
import java.util.Locale;

import io.blackbox_vision.datetimepickeredittext.R;
import io.blackbox_vision.datetimepickeredittext.internal.fragment.DatePickerFragment;
import io.blackbox_vision.datetimepickeredittext.internal.utils.DateUtils;

import static android.view.View.OnFocusChangeListener;
import static android.app.DatePickerDialog.OnDateSetListener;


public final class DatePickerInputEditText extends TextInputEditText implements OnFocusChangeListener, OnDateSetListener {
    private static final String TAG = DatePickerInputEditText.class.getSimpleName();

    private OnFocusChangeListener onFocusChangedListener;
    private FragmentManager manager;
    private String dateFormat;
    private Calendar date;

    public DatePickerInputEditText(Context context) {
        super(context);
        init();
    }

    public DatePickerInputEditText(Context context, AttributeSet attrs) {
        super(context, attrs);
        handleAttributes(attrs);
        init();
    }

    public DatePickerInputEditText(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        handleAttributes(attrs);
        init();
    }

    private void init() {
        setOnFocusChangeListener(this);
    }

    private void handleAttributes(@NonNull AttributeSet attributeSet) {
        try {
            final TypedArray array = getContext().obtainStyledAttributes(attributeSet, R.styleable.DatePickerEditText);

            dateFormat = array.getString(R.styleable.DatePickerEditText_dateFormat);

            array.recycle();
        } catch (Exception ex) {
            ex.printStackTrace();
        }
    }

    @Override
    public void onFocusChange(View view, boolean isFocused) {
        final InputMethodManager imm = (InputMethodManager) getContext().getSystemService(Context.INPUT_METHOD_SERVICE);
        imm.hideSoftInputFromWindow(getWindowToken(), InputMethodManager.HIDE_IMPLICIT_ONLY);

        if (isFocused) {
            final DatePickerFragment datePickerFragment = new DatePickerFragment();

            if (null != date) {
                datePickerFragment.setDate(date);
            }

            datePickerFragment.setOnDateSetListener(this);
            datePickerFragment.show(manager, TAG);
        }

        if (null != onFocusChangedListener) {
            onFocusChangedListener.onFocusChange(view, isFocused);
        }
    }

    @Override
    public void onDateSet(DatePicker datePicker, int year, int monthOfYear, int dayOfMonth) {
        final Calendar calendar = Calendar.getInstance(Locale.getDefault());

        calendar.set(Calendar.YEAR, year);
        calendar.set(Calendar.MONTH, monthOfYear);
        calendar.set(Calendar.DAY_OF_MONTH, dayOfMonth);

        setText(DateUtils.toDate(calendar.getTime(), dateFormat));
        date = calendar;
    }

    public FragmentManager getManager() {
        return manager;
    }

    public DatePickerInputEditText setManager(@NonNull FragmentManager manager) {
        this.manager = manager;
        return this;
    }

    public Calendar getDate() {
        return date;
    }

    public DatePickerInputEditText setDate(@NonNull Calendar date) {
        this.date = date;
        return this;
    }

    public OnFocusChangeListener getOnFocusChangedListener() {
        return onFocusChangedListener;
    }

    public DatePickerInputEditText setOnFocusChangedListener(View.OnFocusChangeListener onFocusChangedListener) {
        this.onFocusChangedListener = onFocusChangedListener;
        return this;
    }

    public String getDateFormat() {
        return dateFormat;
    }

    public DatePickerInputEditText setDateFormat(String dateFormat) {
        this.dateFormat = dateFormat;
        return this;
    }
}

File: datetimepickeredittext/src/main/java/io/blackbox_vision/datetimepickeredittext/view/DatePickerInputEditText.kt
package io.blackbox_vision.datetimepickeredittext.view

import android.content.Context
import android.support.design.widget.TextInputEditText
import android.support.v4.app.FragmentManager
import android.text.InputType
import android.util.AttributeSet
import android.view.View
import android.view.inputmethod.InputMethodManager
import android.widget.DatePicker

import java.text.DateFormat
import java.util.Calendar
import java.util.Locale

import io.blackbox_vision.datetimepickeredittext.R
import io.blackbox_vision.datetimepickeredittext.internal.fragment.DatePickerFragment
import io.blackbox_vision.datetimepickeredittext.internal.utils.DateUtils

import android.view.View.OnFocusChangeListener
import android.view.View.OnClickListener
import android.app.DatePickerDialog.OnDateSetListener
import android.support.v7.app.AppCompatActivity


class DatePickerInputEditText : TextInputEditText, OnFocusChangeListener, OnClickListener, OnDateSetListener {

    private var onFocusChangedListener: OnFocusChangeListener? = null

    private var manager: FragmentManager? = null

    private var focusCount: Int = 0
    private var themeId: Int? = null

    private var dateFormat: String? = null
    private var minDate: String? = null
    private var maxDate: String? = null

    private var textDateFormat: DateFormat? = null

    private var date: Calendar? = null

    constructor(context: Context) : super(context) {
        init()
    }

    constructor(context: Context, attrs: AttributeSet) : super(context, attrs) {
        handleAttributes(attrs)
        init()
    }

    constructor(context: Context, attrs: AttributeSet, defStyleAttr: Int) : super(context, attrs, defStyleAttr) {
        handleAttributes(attrs)
        init()
    }

    private fun init() {
        inputType = InputType.TYPE_NULL
        onFocusChangeListener = this
        setOnClickListener(this)

        /*Set fragment manager*/
        manager = (context as AppCompatActivity).supportFragmentManager
    }

    private fun handleAttributes(attributeSet: AttributeSet) {
        try {
            val array = context.obtainStyledAttributes(attributeSet, R.styleable.DateTimePickerEditText)

            themeId = array.getResourceId(R.styleable.DateTimePickerEditText_theme, 0)

            dateFormat = array.getString(R.styleable.DateTimePickerEditText_dateFormat)
            minDate = array.getString(R.styleable.DateTimePickerEditText_minDate)
            maxDate = array.getString(R.styleable.DateTimePickerEditText_maxDate)

            array.recycle()
        } catch (ex: Exception) {
            ex.printStackTrace()
        }

    }

    override fun onFocusChange(view: View, isFocused: Boolean) {
        val imm = context.getSystemService(Context.INPUT_METHOD_SERVICE) as InputMethodManager
        imm.hideSoftInputFromWindow(windowToken, 0)
        focusCount++

        if (isFocused && focusCount == 1) {
            showDatePicker()
        }

        if (null != onFocusChangedListener) {
            onFocusChangedListener!!.onFocusChange(view, isFocused)
        }
    }

    override fun onClick(v: View) {
        showDatePicker()
    }

    private fun showDatePicker() {
        val datePickerFragment = DatePickerFragment()

        datePickerFragment.date = date
        datePickerFragment.themeId = themeId
        datePickerFragment.minDate = minDate
        datePickerFragment.maxDate = maxDate
        datePickerFragment.onDateSetListener = this

        datePickerFragment.show(manager!!, TAG)
    }

    override fun onDateSet(datePicker: DatePicker, year: Int, monthOfYear: Int, dayOfMonth: Int) {
        val calendar = Calendar.getInstance(Locale.getDefault())

        calendar.set(Calendar.YEAR, year)
        calendar.set(Calendar.MONTH, monthOfYear)
        calendar.set(Calendar.DAY_OF_MONTH, dayOfMonth)

        if (textDateFormat != null) {
            setText(textDateFormat!!.format(calendar.time))
        } else {
            setText(DateUtils.toDate(calendar.time, dateFormat))
        }
        date = calendar
    }

    fun getManager(): FragmentManager? {
        return manager
    }

/*    fun setManager(manager: FragmentManager): DatePickerInputEditText {
        this.manager = manager
        return this
    }*/

    fun getDate(): Calendar? {
        return date
    }

    fun setDate(date: Calendar): DatePickerInputEditText {
        this.date = date
        return this
    }

    fun getOnFocusChangedListener(): OnFocusChangeListener? {
        return onFocusChangedListener
    }

    fun setOnFocusChangedListener(onFocusChangedListener: OnFocusChangeListener): DatePickerInputEditText {
        this.onFocusChangedListener = onFocusChangedListener
        return this
    }

    fun getDateFormat(): String? {
        return dateFormat
    }

    fun setDateFormat(dateFormat: String): DatePickerInputEditText {
        this.dateFormat = dateFormat
        return this
    }

    fun setDateFormat(format: DateFormat): DatePickerInputEditText {
        this.textDateFormat = format
        return this
    }

    fun getThemeId(): Int? {
        return themeId
    }

    fun setThemeId(themeId: Int?): DatePickerInputEditText {
        this.themeId = themeId
        return this
    }

    fun getMaxDate(): String? {
        return maxDate
    }

    fun setMaxDate(maxDate: String): DatePickerInputEditText {
        this.maxDate = maxDate
        return this
    }

    fun getMinDate(): String? {
        return minDate
    }

    fun setMinDate(minDate: String): DatePickerInputEditText {
        this.minDate = minDate
        return this
    }

    companion object {
        private val TAG = "DatePickerInputEditText"
    }
}